// mysqlx/lib/Collection.js
"use strict";
/**
 * Copyright 2019, Danang Galuh Tegar Prasetyo.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
exports.Collection = void 0;
var lodash_isboolean_1 = __importDefault(require("lodash.isboolean"));
var lodash_isnumber_1 = __importDefault(require("lodash.isnumber"));
var lodash_isstring_1 = __importDefault(require("lodash.isstring"));
var CollectionAdd_1 = __importDefault(require("./CollectionAdd"));
var CollectionFind_1 = __importDefault(require("./CollectionFind"));
var CollectionModify_1 = __importDefault(require("./CollectionModify"));
var CollectionRemove_1 = __importDefault(require("./CollectionRemove"));
var OperationResult_1 = __importDefault(require("./OperationResult"));
var Collection = /** @class */ (function () {
    function Collection(schema, xCollection) {
        this.schema = schema;
        this.xCollection = xCollection;
    }
    Collection.createConditionString = function (condition) {
        var conditionString = '';
        if (lodash_isstring_1["default"](condition)) {
            conditionString = condition;
        }
        else if (lodash_isboolean_1["default"](condition)) {
            conditionString = condition.toString();
        }
        else {
            var firstCondition = true;
            for (var key in condition) {
                if (condition.hasOwnProperty(key)) {
                    var value = lodash_isnumber_1["default"](condition[key]) || lodash_isboolean_1["default"](condition[key]) ? condition[key].toString() : "\"" + condition[key] + "\"";
                    conditionString = conditionString
                        .concat(firstCondition ? '' : '  & ')
                        .concat(key)
                        .concat('=')
                        .concat(value);
                    if (firstCondition) {
                        firstCondition = false;
                    }
                }
            }
        }
        return conditionString;
    };
    Collection.prototype.getClient = function () {
        return this.schema.getClient();
    };
    Collection.prototype.getSession = function () {
        return this.schema.getSession();
    };
    Collection.prototype.getSchema = function () {
        return this.schema;
    };
    Collection.prototype.getXCollection = function () {
        return this.xCollection;
    };
    Collection.prototype.add = function (document) {
        try {
            var xCollectionAdd = this.xCollection.add(document);
            return new CollectionAdd_1["default"](this, xCollectionAdd);
        }
        catch (error) {
            throw error;
        }
    };
    Collection.prototype.addOrReplaceOne = function (id, document) {
        return __awaiter(this, void 0, void 0, function () {
            var xResult, error_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.xCollection.addOrReplaceOne(id, document)];
                    case 1:
                        xResult = _a.sent();
                        return [2 /*return*/, new OperationResult_1["default"](xResult)];
                    case 2:
                        error_1 = _a.sent();
                        throw error_1;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    Collection.prototype.count = function () {
        return __awaiter(this, void 0, void 0, function () {
            var error_2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.xCollection.count()];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2:
                        error_2 = _a.sent();
                        throw error_2;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    Collection.prototype.createIndex = function (name, constraint) {
        return __awaiter(this, void 0, void 0, function () {
            var error_3;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.xCollection.createIndex(name, constraint)];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2:
                        error_3 = _a.sent();
                        throw error_3;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    Collection.prototype.dropIndex = function (name) {
        return __awaiter(this, void 0, void 0, function () {
            var error_4;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.xCollection.dropIndex(name)];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2:
                        error_4 = _a.sent();
                        throw error_4;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    Collection.prototype.existsInDatabase = function () {
        return __awaiter(this, void 0, void 0, function () {
            var error_5;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.xCollection.existsInDatabase()];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2:
                        error_5 = _a.sent();
                        throw error_5;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    Collection.prototype.find = function (condition) {
        if (condition === void 0) { condition = true; }
        try {
            var conditionString = Collection.createConditionString(condition);
            var xCollectionFind = this.xCollection.find(conditionString);
            return new CollectionFind_1["default"](this, xCollectionFind);
        }
        catch (error) {
            throw error;
        }
    };
    Collection.prototype.findOne = function (condition) {
        return __awaiter(this, void 0, void 0, function () {
            var findResult, documents, error_6;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.find(condition).execute()];
                    case 1:
                        findResult = _a.sent();
                        documents = findResult.getDocuments();
                        return [2 /*return*/, documents.length > 0 ? documents[0] : void 0];
                    case 2:
                        error_6 = _a.sent();
                        throw error_6;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    Collection.prototype.findByID = function (id) {
        return __awaiter(this, void 0, void 0, function () {
            var error_7;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.findOne({ _id: id })];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2:
                        error_7 = _a.sent();
                        throw error_7;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    Collection.prototype.getName = function () {
        return this.xCollection.getName();
    };
    Collection.prototype.getOne = function (id) {
        return __awaiter(this, void 0, void 0, function () {
            var error_8;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.xCollection.getOne(id)];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2:
                        error_8 = _a.sent();
                        throw error_8;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    Collection.prototype.inspect = function () {
        return this.xCollection.inspect();
    };
    Collection.prototype.modify = function (condition) {
        try {
            var conditionString = Collection.createConditionString(condition);
            var xCollectionModify = this.xCollection.modify(conditionString);
            return new CollectionModify_1["default"](this, xCollectionModify);
        }
        catch (error) {
            throw error;
        }
    };
    Collection.prototype.remove = function (condition) {
        if (condition === void 0) { condition = true; }
        try {
            var conditionString = Collection.createConditionString(condition);
            var xCollectionRemove = this.xCollection.remove(conditionString);
            return new CollectionRemove_1["default"](this, xCollectionRemove);
        }
        catch (error) {
            throw error;
        }
    };
    Collection.prototype.removeByID = function (id) {
        return __awaiter(this, void 0, void 0, function () {
            var error_9;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.remove({ _id: id }).execute()];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2:
                        error_9 = _a.sent();
                        throw error_9;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    Collection.prototype.replaceOne = function (id, document) {
        return __awaiter(this, void 0, void 0, function () {
            var xResult, error_10;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.xCollection.replaceOne(id, document)];
                    case 1:
                        xResult = _a.sent();
                        return [2 /*return*/, new OperationResult_1["default"](xResult)];
                    case 2:
                        error_10 = _a.sent();
                        throw error_10;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    return Collection;
}());
exports.Collection = Collection;
exports["default"] = Collection;
