// mysqlx/lib/OperationResult.js
"use strict";
/**
 * Copyright 2019, Danang Galuh Tegar Prasetyo.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
exports.__esModule = true;
exports.OperationResult = void 0;
var OperationResult = /** @class */ (function () {
    function OperationResult(xResult) {
        this.xResult = xResult;
    }
    OperationResult.prototype.getAffectedItemsCount = function () {
        return this.xResult.getAffectedItemsCount();
    };
    OperationResult.prototype.getAffectedRowsCount = function () {
        return this.xResult.getAffectedRowsCount();
    };
    OperationResult.prototype.getAutoIncrementValue = function () {
        return this.xResult.getAutoIncrementValue();
    };
    OperationResult.prototype.getGeneratedIds = function () {
        return this.xResult.getGeneratedIds();
    };
    OperationResult.prototype.getWarnings = function () {
        return this.xResult.getWarnings();
    };
    OperationResult.prototype.getWarningsCount = function () {
        return this.xResult.getWarningsCount();
    };
    return OperationResult;
}());
exports.OperationResult = OperationResult;
exports["default"] = OperationResult;
