// mysqlx/lib/Schema.js
"use strict";
/**
 * Copyright 2019, Danang Galuh Tegar Prasetyo.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
exports.Schema = void 0;
var Collection_1 = __importDefault(require("./Collection"));
var Table_1 = __importDefault(require("./Table"));
var Schema = /** @class */ (function () {
    function Schema(session, xSchema) {
        this.client = session.getClient();
        this.session = session;
        this.xSchema = xSchema;
    }
    Schema.prototype.getClient = function () {
        return this.client;
    };
    Schema.prototype.getSession = function () {
        return this.session;
    };
    Schema.prototype.getXSchema = function () {
        return this.xSchema;
    };
    Schema.prototype.createCollection = function (name, options) {
        return __awaiter(this, void 0, void 0, function () {
            var xCollection, error_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.xSchema.createCollection(name, options)];
                    case 1:
                        xCollection = _a.sent();
                        return [2 /*return*/, new Collection_1["default"](this, xCollection)];
                    case 2:
                        error_1 = _a.sent();
                        throw error_1;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    Schema.prototype.dropCollection = function (name) {
        return __awaiter(this, void 0, void 0, function () {
            var error_2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.xSchema.dropCollection(name)];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2:
                        error_2 = _a.sent();
                        throw error_2;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    Schema.prototype.existsInDatabase = function () {
        return __awaiter(this, void 0, void 0, function () {
            var error_3;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.xSchema.existsInDatabase()];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2:
                        error_3 = _a.sent();
                        throw error_3;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    Schema.prototype.getCollection = function (name) {
        try {
            var xCollection = this.xSchema.getCollection(name);
            return new Collection_1["default"](this, xCollection);
        }
        catch (error) {
            throw error;
        }
    };
    Schema.prototype.getCollectionAsTable = function (name) {
        try {
            var xTable = this.xSchema.getCollectionAsTable(name);
            return new Table_1["default"](this, xTable);
        }
        catch (error) {
            throw error;
        }
    };
    Schema.prototype.getCollections = function () {
        return __awaiter(this, void 0, void 0, function () {
            var xCollections, collections, _i, xCollections_1, xCollection, error_4;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.xSchema.getCollections()];
                    case 1:
                        xCollections = _a.sent();
                        collections = [];
                        for (_i = 0, xCollections_1 = xCollections; _i < xCollections_1.length; _i++) {
                            xCollection = xCollections_1[_i];
                            collections.push(new Collection_1["default"](this, xCollection));
                        }
                        return [2 /*return*/, collections];
                    case 2:
                        error_4 = _a.sent();
                        throw error_4;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    Schema.prototype.getTable = function (name) {
        try {
            var xTable = this.xSchema.getTable(name);
            return new Table_1["default"](this, xTable);
        }
        catch (error) {
            throw error;
        }
    };
    Schema.prototype.getTables = function () {
        return __awaiter(this, void 0, void 0, function () {
            var xTables, tables, _i, xTables_1, xTable, error_5;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.xSchema.getTables()];
                    case 1:
                        xTables = _a.sent();
                        tables = [];
                        for (_i = 0, xTables_1 = xTables; _i < xTables_1.length; _i++) {
                            xTable = xTables_1[_i];
                            tables.push(new Table_1["default"](this, xTable));
                        }
                        return [2 /*return*/, tables];
                    case 2:
                        error_5 = _a.sent();
                        throw error_5;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    Schema.prototype.getName = function () {
        return this.xSchema.getName();
    };
    Schema.prototype.inspect = function () {
        return this.xSchema.inspect();
    };
    return Schema;
}());
exports.Schema = Schema;
exports["default"] = Schema;
